### PR TITLE
Remove migration lock timeout

### LINKED ISSUE #17578: Execution strategy for migration
Dear all,

I am wondering if the SQL execution strategy specified when configuring the options for the context also covers the Migrate command when it takes place as part of the chain of events of the IWebHost as in the code sample in this [discussion](https://github.com/aspnet/EntityFrameworkCore/issues/9033#issuecomment-317063370)

### LINKED ISSUE #34196: Remove Migrate lock timeout
Also, log when `Migrate` is waiting for a lock

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Log migration lock acquisition
Move seeding to options

Fixes #34196
Fixes #17578

### CODE REVIEW

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 1]
```diff
@@ -701,6 +701,10 @@
     <value>Created transaction savepoint.</value>
     <comment>Debug RelationalEventId.CreatedTransactionSavepoint</comment>
   </data>
+  <data name="LogCreatingMigrationLock" xml:space="preserve">
+    <value>Acquiring an exclusive lock for migration application. See https://aka.ms/efcore-docs-migrations for more information if this takes too long.</value>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:705]: @ajcvickers Could you create a more specific link for this?

**Comment by roji** [comment]
See comment on possibly continuing to have cancellation and timeout, but fully client-managed: https://github.com/dotnet/efcore/issues/34196#issuecomment-2266650727

**Comment by AndriySvyryd** [comment]
Follow-up from the API review: using `UseSeeding` on the generic `DbContextOptionsBuilder<>` doesn't produce ambiguity. `Action<DbContext, bool>` is chosen by default and `Action<TContext, bool>` is chosen if the lambda has code referencing anything a member on `TContext` that's not on `DbContext`.

TLDR: it just works!

File: src/EFCore.Relational/Migrations/Internal/MigrationCommandExecutor.cs [hunk 2]
```diff
@@ -36,19 +29,20 @@ public MigrationCommandExecutor(IExecutionStrategy executionStrategy)
         IEnumerable<MigrationCommand> migrationCommands,
         IRelationalConnection connection)
     {
+        var commands = migrationCommands.ToList();
```
Comment (roji) [inline src/EFCore.Relational/Migrations/Internal/MigrationCommandExecutor.cs:32]: Definitely should consider changing the parameter to be a list (I know we discussed already on the old PR, not sure what the outcome was)

File: src/EFCore.Relational/Migrations/Internal/Migrator.cs [hunk 3]
```diff
@@ -94,12 +92,12 @@ public virtual void Migrate(Action<DbContext, IMigratorData>? seed, string? targ
         {
             _databaseCreator.Create();
         }
-
```
Comment (roji) [inline src/EFCore.Relational/Migrations/Internal/Migrator.cs:97]: I kinda liked this empty line

File: src/EFCore.Relational/Migrations/Internal/Migrator.cs [hunk 4]
```diff
@@ -111,28 +109,30 @@ public virtual void Migrate(Action<DbContext, IMigratorData>? seed, string? targ
...
 
-            foreach (var plugin in _plugins)
-            {
-                plugin.Migrated(_currentContext.Context, migratorData);
-            }
+            var coreOptionsExtension =
+                _contextOptions.FindExtension<CoreOptionsExtension>()
+                ?? new CoreOptionsExtension();
 
+            var seed = coreOptionsExtension.Seeder;
```
Comment (roji) [inline src/EFCore.Relational/Migrations/Internal/Migrator.cs:122]: seed or seeder :trollface:

File: src/EFCore.Relational/Properties/RelationalStrings.resx (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Properties/RelationalStrings.resx:705]: @AndriySvyryd you can make one up and we'll create it later

File: src/EFCore.Relational/Storage/RelationalDatabaseCreator.cs [hunk 5]
```diff
@@ -267,29 +284,48 @@ public virtual bool EnsureCreated()
...
+                throw new InvalidOperationException(CoreStrings.MissingSeeder);
             }
         }
         finally
         {
             await transactionScope.DisposeAsyncIfAvailable().ConfigureAwait(false);
         }
 
-        return false;
+        return operationsPerformed;
```
Comment (roji) [inline src/EFCore.Relational/Storage/RelationalDatabaseCreator.cs:328]: Am not too sure about the DisposeAsynfIfAvailable here... TransactionScope doesn't implement DisposeAsync (System.Transactions is generally sync-only), so we should just do a regular using, like in EnsureCreated, no?

File: src/EFCore.Cosmos/Storage/Internal/CosmosDatabaseCreator.cs [hunk 6]
```diff
@@ -55,7 +61,21 @@ public virtual bool EnsureCreated()
 
         if (created)
         {
-            Seed();
+            InsertData();
```
Comment (roji) [inline src/EFCore.Cosmos/Storage/Internal/CosmosDatabaseCreator.cs:64]: Oh, I wasn't aware that we supported HasData() seeding for Cosmos...

**Review by roji — APPROVED** [review]

File: src/EFCore.Relational/Migrations/Internal/MigrationCommandExecutor.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Migrations/Internal/MigrationCommandExecutor.cs:32]: We should remove this `.ToList` eventually, see https://github.com/dotnet/efcore/issues/19710

File: src/EFCore.Relational/Migrations/Internal/Migrator.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Migrations/Internal/Migrator.cs:122]: Yes

File: src/EFCore.Relational/Migrations/Internal/MigrationCommandExecutor.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Relational/Migrations/Internal/MigrationCommandExecutor.cs:33]: OK. As discussed before, the sample scenario is seeding with massive data, which is something that basically doesn't work for various other reasons (see https://github.com/dotnet/efcore/issues/19710#issuecomment-2287973850).
